src/ocean/math/BinaryHistogram.d

```diff
@@ -0,0 +1,384 @@
+
+        ***********************************************************************/
+
+        private static bool fieldOffsetsCorrect ( )
+        {
+            foreach ( i, field; typeof(Bins.tupleof) )
+                if ( Bins.tupleof[i].offsetof != i * BinaryHistogram.init.bins[i].sizeof )
+                    return false;
+            return true;
+        }
+
+        /***********************************************************************
+
+            CTFE generator of the fields of this struct.
+
+            Params:
+                max_power = maximum power of 2 contained within a bin
+                TODO
+
+            Returns:
+                code for the series of fields for bins up to the specified
+                maximum power of 2. (See unittest for examples.)
+
+        ***********************************************************************/
+
+        private static istring divisionBinVariables ( uint max_power,
+            istring suffix )
+        {
+            const type = typeof(BinaryHistogram.bins[0]).stringof;
+
+            istring res;
+
+            istring formatCount ( ulong count )
+            {
+                const prefixes = [""[], "Ki", "Mi", "Gi", "Ti", "Pi", "Ei"];
+
+                uint prefix;
+                while ( count >= 1024 )
+                {
+                    count /= 1024;
+                    prefix++;
+                }
+                return ctfe_i2a(count) ~ prefixes[prefix] ~ suffix;
+            }
+
+            res ~= type ~ " from_" ~ formatCount(0) ~ ";";
+
+            for ( size_t power = 0; power <= max_power; power++ )
+                res ~= type ~ " from_" ~ formatCount(1 << power) ~ ";";
+
+            return res;
+        }
+
+        unittest
+        {
+            test!("==")(divisionBinVariables(0, "B"), "uint from_0B;uint from_1B;");
+            test!("==")(divisionBinVariables(16, "B"),
+                "uint from_0B;uint from_1B;uint from_2B;uint from_4B;uint from_8B;uint from_16B;uint from_32B;uint from_64B;uint from_128B;uint from_256B;uint from_512B;uint from_1KiB;uint from_2KiB;uint from_4KiB;uint from_8KiB;uint from_16KiB;uint from_32KiB;uint from_64KiB;");
+        }
+
+        /// Fields.
+        mixin(divisionBinVariables(MaxPow2, Suffix));
+    }
+
+    /// The number of fields of Bins must equal the length of the fixed-length
+    /// array this.bins.
+    static assert(Bins.tupleof.length == bins.length);
+
+    /***************************************************************************
+
+        Counts the specified value by incrementing the corresponding histogram
+        bin and the total number of transactions and adding `n` to the total
+        count.
+
+        Params:
+            n = the value to add to the histogram
+
+        Returns:
+            n
+
+    ***************************************************************************/
+
+    public ulong record ( ulong n )
```

Comment:
Looking again, I'm not sure about this method name. "Record" is a verb and a noun, so I think it's a bit confusing. Maybe simply `add`?

Reply:
Good idea, or `count`?

Reply:
`opAddAssign` :trollface: